File: JAVA/LetterCombinationsofaPhoneNumber.java
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
	static HashMap<Character, String> map = new HashMap<Character, String>();
	static{
		map.put('2', "abc");
		map.put('3', "def");
		map.put('4', "ghi");
		map.put('5', "jkl");
		map.put('6', "mno");
		map.put('7', "pqrs");
		map.put('8', "tuv");
		map.put('9', "wxyz");
	}
	ArrayList<String> ret = new ArrayList<String>();
    public ArrayList<String> letterCombinations(String digits) {
    	ret.clear();
        if(digits == null || digits.length() == 0)
        {
        	ret.add("");
        	return ret;
        }
        StringBuilder sb = new StringBuilder();
        dfs(digits, 0, sb);
        return ret;
    }
	private void dfs(String digits, int start, StringBuilder sb) {
		if(start >= digits.length())
		{
			ret.add(new String(sb.toString()));
			return;
		}
		String cur = map.get(digits.charAt(start));
		{
			for(int i = 0; i < cur.length(); i++)
			{
				sb.append(cur.charAt(i));
				dfs(digits, start+1, sb);
				sb.deleteCharAt(sb.length()-1);
			}
		}
	}
}
public class LetterCombinationsofaPhoneNumber {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C++/LetterCombinationsofaPhoneNumber.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
#include<map>
using namespace std;
map<char, string> vm;
void init()
{
	vm['2'] = "abc";
	vm['3'] = "def";
	vm['4'] = "ghi";
	vm['5'] = "jkl";
	vm['6'] = "mno";
	vm['7'] = "pqrs";
	vm['8'] = "tuv";
	vm['9'] = "wxyz";
}
class Solution {
public:
	vector<string> ret;
	void dfs(string digits, int index, int size, string& list)
	{
		if(index == size)
		{
			ret.push_back(list);
			return;
		}
		string alpha = vm.find(digits[index])->second;
		for(int i = 0; i < alpha.size(); i++)
		{
			list.push_back(alpha[i]);
			dfs(digits, index+1, size, list);
			list.pop_back();
		}
	}
    vector<string> letterCombinations(string digits) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		init();
		ret.clear();
		string list;
		dfs(digits, 0, digits.size(), list);
        return ret;
    }
};
int main()
{
	string d = "2";
	Solution sol;
	sol.letterCombinations(d);
	getchar();
}